lib/debug/repl.cpp

```diff
@@ -0,0 +1,211 @@
+    if (!adc.HasMember("pins")) {
+      log_.log(hyped::core::LogLevel::kFatal,
+               "Missing required field 'io.adc.pins' in configuration file");
+      return std::nullopt;
+    }
+    const auto pins = adc["pins"].GetArray();
+    for (auto &pin : pins) {
+      repl->addAdcCommands(pin.GetUint());
+    }
+  }
+
+  if (!io.HasMember("i2c")) {
+    log_.log(hyped::core::LogLevel::kFatal,
+             "Missing required field 'io.i2c' in configuration file");
+    return std::nullopt;
+  }
+  const auto i2c = io["i2c"].GetObject();
+  if (!i2c.HasMember("enabled")) {
+    log_.log(hyped::core::LogLevel::kFatal,
+             "Missing required field 'io.i2c.enabled' in configuration file");
+    return std::nullopt;
+  }
+  if (i2c["enabled"].GetBool()) {
+    if (!i2c.HasMember("channels")) {
+      log_.log(hyped::core::LogLevel::kFatal,
+               "Missing required field 'io.i2c.channels' in configuration file");
+      return std::nullopt;
+    }
+    const auto channels = i2c["channels"].GetArray();
+    for (auto &channel : channels) {
+      repl->addI2cCommands(channel.GetUint());
+    }
+  }
+  return repl;
+}
+
+void Repl::printCommands()
+{
+  log_.log(hyped::core::LogLevel::kInfo, "Available commands:");
+  for (const auto &[name, command] : command_map_) {
+    log_.log(hyped::core::LogLevel::kInfo, "  %s: %s", name.c_str(), command.description.c_str());
+  }
+}
+
+void Repl::handleCommand()
+{
+  std::cout << "> ";
+  std::string command;
+  std::getline(std::cin, command);
+  auto nameAndCommand = command_map_.find(command);
+  if (nameAndCommand == command_map_.end()) {
+    log_.log(hyped::core::LogLevel::kFatal, "Unknown command: %s", command.c_str());
+    return;
+  }
+  nameAndCommand->second.handler();
+}
+
+void Repl::addCommand(const Command &cmd)
+{
+  command_map_.emplace(cmd.name, cmd);
+  log_.log(hyped::core::LogLevel::kDebug, "Added command: %s", cmd.name.c_str());
+}
+
+void Repl::addQuitCommand()
+{
+  addCommand(Command{"quit", "Quit the REPL", [this]() { exit(0); }});
+}
+
+void Repl::addHelpCommand()
+{
+  addCommand(Command{"help", "Print this help message", [this]() { printCommands(); }});
+}
+
+void Repl::addAdcCommands(const uint8_t pin)
+{
+  const auto adc = std::make_shared<hyped::io::Adc>(pin, log_);
+  Command adc_read_command;
+  std::stringstream identifier;
+  identifier << "adc " << static_cast<int>(pin) << " read";
+  adc_read_command.name = identifier.str();
+  std::stringstream description;
+  description << "Read from ADC pin " << static_cast<int>(pin);
+  adc_read_command.description = description.str();
+  adc_read_command.handler     = [this, adc]() {
+    const auto value = adc->readValue();
+    if (value) { log_.log(hyped::core::LogLevel::kInfo, "ADC value: %d", *value); }
+  };
+  addCommand(adc_read_command);
+}
+
+void Repl::addI2cCommands(const uint8_t channel)
+{
+  const auto i2c = std::make_shared<hyped::io::I2c>(channel, log_);
+  {
+    Command i2c_read_command;
```

Comment:
not a change rn, just an fyi: there will have to be a separate readReg and read command once my pr is merged

Reply:
Sure, we can add commands as the interface changes.